overlord/devicestate/handlers_remodel.go

```diff
@@ -151,6 +178,100 @@ func (m *DeviceManager) doSetModel(t *state.Task, _ *tomb.Tomb) (err error) {
 	return nil
 }
 
+func trackedValidationSetsFromModel(st *state.State, model *asserts.Model) ([]*asserts.ValidationSet, error) {
+	currentSets, err := assertstate.TrackedEnforcedValidationSets(st)
+	if err != nil {
+		return nil, err
+	}
+
+	var fromModel []*asserts.ValidationSet
+	for _, mvs := range model.ValidationSets() {
+		for _, cvs := range currentSets.Sets() {
+			if mvs.SequenceName() == cvs.SequenceName() {
+				fromModel = append(fromModel, cvs)
+			}
+		}
+	}
+	return fromModel, nil
+}
+
+func rollBackValidationSets(st *state.State, oldSets []*asserts.ValidationSet, newSets []*asserts.ModelValidationSet, deviceCtx snapstate.DeviceContext) error {
+	for _, set := range newSets {
+		if err := assertstate.ForgetValidationSet(st, set.AccountID, set.Name, assertstate.ForgetValidationSetOpts{
+			ForceForget: true,
+		}); err != nil {
+			fmt.Println("we did not roll them back!", err)
+			return err
+		}
+	}
+
+	vSetKeys := make(map[string][]string, len(oldSets))
+	for _, vs := range oldSets {
+		sequenceName := vs.SequenceName()
+		vSetKeys[sequenceName] = vs.At().PrimaryKey
+	}
+
+	snaps, ignore, err := snapstate.InstalledSnaps(st)
+	if err != nil {
+		return err
+	}
+
+	// TODO: is this right? i think we have to ignore everything that is
+	// installed, since we won't have run the undos for the newly installed
+	// snaps yet
+	for _, sn := range snaps {
+		ignore[sn.SnapName()] = true
+	}
+
+	if err := assertstate.ApplyLocalEnforcedValidationSets(st, vSetKeys, nil, snaps, ignore); err != nil {
+		return err
+	}
```

Comment:
Yes, at this stage it's a bit hard to retrack old validation-sets as the undo for snaps have not run, meaning that snaps might have the new revisions or have new snaps installed. The only other alternative would be to split up the undo logic for rolling back validation-sets, so the undo task would be at the early stage of the change with a `nil` do logic, and then having a `nil` undo logic for a `update-validation-sets` task. However the undo logic would then need to ignore if the new validation-sets haven't been tracked yet.

In any case this needs a more descriptive comment (explaining why this choice is made and maybe highlighting the alternative) and remove the TODO